dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java

```diff
@@ -831,32 +843,50 @@ protected String getItemsSql( EventQueryParams params, SqlHelper hlp )
     {
         // Creates a map grouping queryItems referring to repeatable stages and
         // those referring to non-repeatable stages
+        // Only for enrollments, for events all query items are treated as
+        // non-repeatable
         Map<Boolean, List<QueryItem>> itemsByRepeatableFlag = params.getItems()
             .stream()
             .filter( QueryItem::hasFilter )
-            .collect( Collectors.groupingBy(
-                QueryItem::hasRepeatableStageParams ) );
+            .collect( groupingBy(
+                queryItem -> queryItem.hasRepeatableStageParams() && params.getEndpointItem() == ENROLLMENT ) );
+
+        // groups repeatable conditions based on PSID+DEID
+        Map<String, List<String>> repeatableConditionsByIdentifier = asSqlCollection( itemsByRepeatableFlag.get( true ),
+            params )
+                .collect( groupingBy(
+                    IdentifiableSql::getIdentifier,
+                    mapping( IdentifiableSql::getSql, toList() ) ) );
+
+        // joins each group with OR
+        Collection<String> orConditions = repeatableConditionsByIdentifier.values()
+            .stream()
+            .map( sameGroup -> joinSql( sameGroup, OR_JOINER ) )
+            .collect( toList() );
 
-        Collection<String> repeatableSqlConditions = asSqlCollection( itemsByRepeatableFlag.get( true ), params );
-        Collection<String> nonRepeatableSqlConditions = asSqlCollection( itemsByRepeatableFlag.get( false ), params );
+        // non repeatable conditions
+        Collection<String> andConditions = asSqlCollection( itemsByRepeatableFlag.get( false ), params )
+            .map( IdentifiableSql::getSql )
+            .collect( toList() );
 
-        if ( repeatableSqlConditions.isEmpty() && nonRepeatableSqlConditions.isEmpty() )
+        if ( orConditions.isEmpty() && andConditions.isEmpty() )
         {
             return "";
         }
 
-        String repeatableSql = joinSql( repeatableSqlConditions, joining( " or ", "(", ")" ) );
-        String nonRepeatableSql = joinSql( nonRepeatableSqlConditions, joining( _AND_ ) );
+        return hlp.whereAnd() + " " + joinSql( Stream.concat(
```

Comment:
What happens if `orConditions` is empty or `andConditions` is empty?
Could it result in something like `and "" and element=1`

Reply:
Actually not, I just noticed the `Stream.concat` thing =D